internal/impl/aws/input_sqs.go

```diff
@@ -181,99 +199,116 @@ func (a *awsSQSReader) Connect(ctx context.Context) error {
+		fifo:    list.New(),
+		limit:   a.conf.MaxOutstanding,
+		timeout: a.conf.MessageTimeout,
 	}
+	ift.l = sync.NewCond(&ift.m)
 
 	var wg sync.WaitGroup
-	wg.Add(2)
+	wg.Add(3)
 	go a.readLoop(&wg, ift)
 	go a.ackLoop(&wg, ift)
+	go a.refreshLoop(&wg, ift)
 	go func() {
 		wg.Wait()
 		a.closeSignal.TriggerHasStopped()
 	}()
 	return nil
 }
 
-type sqsInFlightHandle struct {
-	receiptHandle  string
-	timeoutSeconds int
-	addedAt        time.Time
-}
-
 type sqsInFlightTracker struct {
-	handles map[string]sqsInFlightHandle
+	handles map[string]*list.Element
+	fifo    *list.List // contains *sqsMessageHandle
+	limit   int
+	timeout time.Duration
 	m       sync.Mutex
+	l       *sync.Cond
 }
 
-func (t *sqsInFlightTracker) PullToRefresh() (handles []sqsMessageHandle, timeoutSeconds int) {
+func (t *sqsInFlightTracker) PullToRefresh(limit int) []*sqsMessageHandle {
 	t.m.Lock()
 	defer t.m.Unlock()
 
-	handles = make([]sqsMessageHandle, 0, len(t.handles))
-	for k, v := range t.handles {
-		if time.Since(v.addedAt) < time.Second {
-			continue
-		}
-		handles = append(handles, sqsMessageHandle{
-			id:            k,
-			receiptHandle: v.receiptHandle,
-		})
-		if v.timeoutSeconds > timeoutSeconds {
-			timeoutSeconds = v.timeoutSeconds
+	handles := make([]*sqsMessageHandle, 0, limit)
+	now := time.Now()
+	// Pull the front of our fifo until we reach our limit or we reach elements that do not
+	// need to be refreshed
+	for e := t.fifo.Front(); e != nil && len(handles) < limit; e = t.fifo.Front() {
+		v := e.Value.(*sqsMessageHandle)
+		if v.deadline.Sub(now) > (t.timeout / 2) {
+			break
 		}
+		handles = append(handles, v)
+		v.deadline = now.Add(t.timeout)
+		// Keep our fifo in deadline sorted order
+		t.fifo.MoveToBack(e)
 	}
-	return
+	return handles
+}
+
+func (t *sqsInFlightTracker) Size() int {
+	t.m.Lock()
+	defer t.m.Unlock()
+	return len(t.handles)
 }
 
 func (t *sqsInFlightTracker) Remove(id string) {
 	t.m.Lock()
 	defer t.m.Unlock()
-	delete(t.handles, id)
+	entry, ok := t.handles[id]
+	if ok {
+		t.fifo.Remove(entry)
+		delete(t.handles, id)
+	}
+	t.l.Signal()
 }
 
-func (t *sqsInFlightTracker) AddNew(messages ...types.Message) {
+func (t *sqsInFlightTracker) IsTracking(id string) bool {
 	t.m.Lock()
 	defer t.m.Unlock()
+	_, ok := t.handles[id]
+	return ok
+}
 
-	for _, m := range messages {
-		if m.MessageId == nil || m.ReceiptHandle == nil {
-			continue
-		}
+func (t *sqsInFlightTracker) Clear() {
+	t.m.Lock()
+	defer t.m.Unlock()
+	clear(t.handles)
+	t.fifo = list.New()
+	t.l.Signal()
+}
 
-		handle := sqsInFlightHandle{
-			timeoutSeconds: 30,
-			receiptHandle:  *m.ReceiptHandle,
-			addedAt:        time.Now(),
-		}
-		if timeoutStr, exists := m.Attributes[sqsiAttributeNameVisibilityTimeout]; exists {
-			// Might as well keep the queue timeout setting refreshed as we
-			// consume new data.
-			if tmpTimeoutSeconds, err := strconv.Atoi(timeoutStr); err == nil {
-				handle.timeoutSeconds = tmpTimeoutSeconds
-			}
```

Comment:
FYI this attribute is not mentioned in the docs and doesn't seem to be present in my testing...